Persistence class to save and restore models

dmonllao opened the pull request:
Persistence based on PHP native Serializable interface, classes interested in being backed up and restored only need to implement the native interface.

I wanted to follow http://scikit-learn.org/stable/modules/model_persistence.html approach (instead of a separate Persistence class) but given that PHP is basically a web language I imagine people being interested in adding a stuff like \Phpml\Persistence::saveToDatabase in future and may be better to isolate the save/restore methods from model classes. I also though of an extra interface, but having \Serializable it may seem redundant or confusing for people.

I've only added support for KNearestNeighbors, but I am keen to add a docs page and expand the support to other classifiers if this approach is accepted.

I dislike that the Serializable interface introduces another dodgy way to cheat and get read and write access to private and protected attributes; people can always use reflection anyway so it does not introduce any new issue...

Related to #29

itcraftsmanpl commented on src/Phpml/Classification/KNearestNeighbors.php:
```diff
@@ -9,7 +9,7 @@
 use Phpml\Math\Distance\Euclidean;
 
-class KNearestNeighbors implements Classifier
+class KNearestNeighbors implements Classifier, \Serializable
```
Instead of `Serializable` interface you could use `__sleep()` and `__wakeup()`.

itcraftsmanpl commented on src/Phpml/Persistence.php:
```diff
@@ -0,0 +1,46 @@
+use Phpml\Exception\PersistenceException;
+use Phpml\Exception\FileException;
+
+class Persistence
```
I like the idea, but can we work on the name. Maybe `ModelManager` or `ModleKeeper` or something else?

itcraftsmanpl commented on src/Phpml/Persistence.php:
```diff
@@ -0,0 +1,46 @@
+     * @param \Serializable $object
+     * @param string        $filepath
+     */
+    public static function saveToFile(\Serializable $object, string $filepath)
```
We could typehint `Estimator` and make sure that only estimators are serialized.

itcraftsmanpl commented on src/Phpml/Persistence.php:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @return \Serializable
+     */
+    public static function restoreFromFile(string $filepath)
```
return type `Estimator`

itcraftsmanpl commented on tests/Phpml/PersistenceTest.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $filepath = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'unexisting';
+
+        $obj = new \SplObjectStorage();
```
Not needed.

itcraftsmanpl commented:
@dmonllao great job. I started working on same functionality, but you were faster.

dmonllao commented:
Thanks for the review @itcraftsmanpl.

I've updated the patch. I didn't expected the default PHP serialization to work that well, I've switched KNearestNeighbors to the default objects serialization and it worked out of the box, same for all other estimators. Serializable interface is quite more powerful than __sleep and __wakeup, in this project's context I can only think of __sleep being used to discard unnecessary stuff that should not be serialized. In any case, with this patch in it will be up to the estimator to use __sleep/__wakeup or Serializable

Other changes introduced:
* Persistence renamed to ModelManager
* ModelManager methods type hinting Estimator classes 
* Tests for all estimators. Under each estimator as they are responsible of preparing their data properly to be serialized 
* New docs page

itcraftsmanpl commented:
@dmonllao impressive :+1: :+1: :+1:

itcraftsmanpl commented on src/Phpml/ModelManager.php:
```diff
@@ -0,0 +1,52 @@
+     * @param Estimator     $object
+     * @param string        $filepath
+     */
+    public static function saveToFile(Estimator $object, string $filepath)
```
I don't like static calls, do you have better idea? Or maybe this is the only way?

itcraftsmanpl replied:
Maybe we could add some kind of abstraction for storage layer (like flysystem). And then:
```
$manager = new ModelManager($fileSystem);
$manager->saveToFile();
```

dmonllao replied:
Sorry, I don't have time to implement abstracted storage. Also note that it may not be a good idea to bind a models manager entity to files management (I'm talking about your suggestion above) I am not a fan of static stuff either, I guess we could do something like the snipped pasted below, but I see more boilerplate code and not much benefit (testing is not affected by these methods being static) Anyway, if you really want to avoid statics I can update the patch with something like this:

`
$manager = new ModelManager();
$manager->saveToFile($object, $filepath);
`

itcraftsmanpl replied:
Ok, we can leave it static. I don't have better idea for now :)

itcraftsmanpl replied:
Or maybe not, if leave it static and later change to object then we introduce bc break. By removing static call code can be more flexible. @dmonllao who you think about that?

dmonllao replied:
Yeah, let's do it. I've just added an extra commit (same branch) switching to

`
$manager = new ModelManager();
$manager->saveToFile($object, $filepath);
`

I've bulk-replaced all appearances and unit tests are passing, it should be ready to merge.

itcraftsmanpl replied:
:+1:

itcraftsmanpl replied:
Travis have some problems today. I will merg it tomorrow and create new release. Great job :+1:

itcraftsmanpl commented on src/Phpml/ModelManager.php:
```diff
@@ -0,0 +1,52 @@
+     */
+    public function saveToFile(Estimator $object, string $filepath)
+    {
+        if (!file_exists($filepath) || !is_writable(dirname($filepath))) {
```
@dmonllao file must exist first, is this nesesery? Before i can save model i must create new empty file. Do you think is this correct approach?

dmonllao replied:
No this is wrong, I must have f*cked it up at some point, give me 10 mins and I send a patch

itcraftsmanpl replied:
cool, there is no need to hurry :)

dmonllao replied:
I was adding a dataProvider to testSaveToWrongFile and I was playing with some ugly OS-dependant filepaths but I ended up adding that much conditions to the provider code that the test was too tied to the code.

The bug was covered by a dependency between unit tests...

You can probably see the fix below: https://github.com/dmonllao/php-ml/commit/e44bccf3922ad043a2c1100a9820195537b9a408

dmonllao commented:
Hi @itcraftsmanpl, Yesterday I added https://github.com/dmonllao/php-ml/commit/e44bccf3922ad043a2c1100a9820195537b9a408 (persistence-fix branch) not sure if you missed it, the patch fixes the issue you commented about.

I don't know how do you use to merge (CLI or github) let me know if you want me to create a new github pull request for it.
